src/overlays/actors/ovl_En_Rat/z_en_rat.c

```diff
@@ -28,123 +47,913 @@ const ActorInit En_Rat_InitVars = {
+
+    posB.x = this->actor.world.pos.x - (this->axisUp.x * 4.0f);
+    posB.y = this->actor.world.pos.y - (this->axisUp.y * 4.0f);
+    posB.z = this->actor.world.pos.z - (this->axisUp.z * 4.0f);
+
+    if (EnRat_IsOnCollisionPoly(play, &posA, &posB, &posUpDown, &polyUpDown, &bgIdUpDown)) {
+        posB.x = (this->axisForwards.x * lineLength) + posA.x;
+        posB.y = (this->axisForwards.y * lineLength) + posA.y;
+        posB.z = (this->axisForwards.z * lineLength) + posA.z;
+
+        if (EnRat_IsOnCollisionPoly(play, &posA, &posB, &posSide, &polySide, &bgIdSide)) {
+            if ((polySide != NULL) && this->hasLostTrackOfPlayer) {
+                return false;
+            }
+
+            this->shouldRotateOntoSurfaces |= EnRat_UpdateFloorPoly(this, polySide, play);
+            Math_Vec3f_Copy(&this->actor.world.pos, &posSide);
+            this->actor.floorBgId = bgIdSide;
+            this->actor.speedXZ = 0.0f;
+        } else {
+            if (polyUpDown != this->actor.floorPoly) {
+                this->shouldRotateOntoSurfaces |= EnRat_UpdateFloorPoly(this, polyUpDown, play);
+            }
+
+            Math_Vec3f_Copy(&this->actor.world.pos, &posUpDown);
+            this->actor.floorBgId = bgIdUpDown;
+        }
+    } else {
+        this->actor.speedXZ = 0.0f;
+        lineLength *= 3.0f;
+        Math_Vec3f_Copy(&posA, &posB);
+
+        for (i = 0; i < 3; i++) {
+            if (i == 0) {
+                // backwards
+                posB.x = posA.x - (this->axisForwards.x * lineLength);
+                posB.y = posA.y - (this->axisForwards.y * lineLength);
+                posB.z = posA.z - (this->axisForwards.z * lineLength);
+            } else if (i == 1) {
+                // left
+                posB.x = posA.x + (this->axisLeft.x * lineLength);
+                posB.y = posA.y + (this->axisLeft.y * lineLength);
+                posB.z = posA.z + (this->axisLeft.z * lineLength);
+            } else {
+                // right
+                posB.x = posA.x - (this->axisLeft.x * lineLength);
+                posB.y = posA.y - (this->axisLeft.y * lineLength);
+                posB.z = posA.z - (this->axisLeft.z * lineLength);
+            }
+
+            if (EnRat_IsOnCollisionPoly(play, &posA, &posB, &posSide, &polySide, &bgIdSide)) {
+                this->shouldRotateOntoSurfaces |= EnRat_UpdateFloorPoly(this, polySide, play);
+                Math_Vec3f_Copy(&this->actor.world.pos, &posSide);
+                this->actor.floorBgId = bgIdSide;
+                break;
+            }
+        }
+
+        if (i == 3) {
+            // no collision nearby
+            return false;
+        }
+    }
+
+    return true;
+}
+
+/**
+ * Transform coordinates from actor coordinate space to world space, according to current orientation.
+ * `offset` is expected to already be at world scale.
+ */
+void EnRat_ActorCoordsToWorld(EnRat* this, Vec3f* offset, Vec3f* pos) {
```

Comment:
Could you make an note about what actor coordinate space is? It's distinct from model space right? Which is what I originally thought it was.